pkg/enforcement/check.go

```diff
@@ -30,31 +38,90 @@ func NewCheckRequest(user User, action Action, resource Resource, context map[st
 	}
 }
 
+func newJsonCheckRequest(opaUrl string, user User, action Action, resource Resource, context map[string]string) ([]byte, error) {
+	checkReq := NewCheckRequest(user, action, resource, context)
+	var jsonCheckReq []byte
+	var err error
+	if opaUrl != "" {
+		opaCheckReq := &opaCheckRequest{
+			checkReq,
+		}
+		jsonCheckReq, err = json.Marshal(opaCheckReq)
+	} else {
+		jsonCheckReq, err = json.Marshal(checkReq)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return jsonCheckReq, nil
+}
+
+func (e *PermitEnforcer) getCheckEndpoint() string {
+	if opaUrl := e.config.GetOpaUrl(); opaUrl != "" {
+		return opaUrl + "/v1/data/" + mainPolicyPath
+	} else {
+		return e.config.GetPdpUrl() + "/allowed"
+	}
+}
+
+func (e *PermitEnforcer) parseResponse(res *http.Response) (*AuthorizationResult, error) {
+	var result AuthorizationResult
+	var resMap map[string]interface{}
+	err := json.NewDecoder(res.Body).Decode(&resMap)
```

Comment:
To be honest, this function's code feels pretty chaotic and too complex to me.
You basically decode json bytes to a map only to marshal it back to bytes immediately again.

I would suggest the following options (excuse my rusted go, not necessarily working as is):
1.
```
        actualAuthResMap := &resMap
	err := json.NewDecoder(res.Body).Decode(actualAuthResMap)
	if err != nil {
	   ...
	}
        if e.config.GetOpaUrl() != "" { 
            actualAuthResMap = resMap["result"].(map[string]interface{})
        }
        // From here use `mapstructure` to parse the map into `AuthorizationResult`

```
2. Get the bytes (`res.Body`) without decoding. If it's not from OPA - Unmarshal it into `AuthorizationResult`. If it is - Unmarshal it into (you can use an inline struct):
```
opaNestedRes struct {
    inner AuthorizationResult
}
```
3. Search google for how to conditionally parse a json?